### DISCUSSION TITLE
OSS JS bundle/assets code push as a first class citizen of React Native ecosystem

### DISCUSSION THREAD

**Discussion by JofBigHealth** [react-native-community/discussions-and-proposals#365] [Ideas]
### Background

Microsoft provide a service called [Code Push](https://microsoft.github.io/code-push/) to push new JavaScript bundles - and assets - to a React Native app without requiring builds be uploaded to Apple's and Google's stores.

It's an excellent product but unfortunately we cannot use it as it currently stands due to a mismatch between our strict security policies and what Microsoft are willing to offer. Additionally we cannot use Expo - which I gather offers something similar - because our app is incompatible with that platform.

To that end, we would like to host our own "code push" service that has similar features.

### The Core of It

If a code push service/library were developed as a first class citizen of the React Native ecosystem we could ensure:
a) it would solve for us the problems described in Background.
b) critically, it would keep feature parity with RN core which could open up new innovative functionality.
c) it opens up such capabilities to a wider audience than can currently use Code Push.
d) it would enable faster iteration for all developers and in the process probably open up innovation in RN continuous deployment/integration.
e) React Native itself would benefit from (d) because faster iteration means faster discovery and fixing of React Native core bugs.
f) IDE and IDE plugin developers could integrate with this functionality, knowing they don't have to rely on a third party commercial product.

### Discussion points

1) Is this viable?
2) Are there already any OSS projects that solve this problem? (I admit I'm not super up to date with what's available)

**Comment by brentvatne** [comment]
@JofBigHealth - fwiw we are currently refactoring the Expo updates API to make it compatible with any React Native app

**Comment by JofBigHealth** [comment]
@brentvatne Amazing! That's a very generous contribution to the community, thank you.
